### PR TITLE
feat(rpc): rpc handler for eth_getUncleByBlock{Hash/Number}AndIndex

### PR DESCRIPTION
**Pull request by fkrause98**
## Changes:
- Added handler for eth_getUncleByBlockHashAndIndex and eth_getUncleByBlockNumberAndIndex.
- Related: #1225.

### CODE REVIEW

File: crates/rpc/rpc/src/eth/api/block.rs [hunk 1]
```diff
@@ -23,6 +23,19 @@ where
...
 
+    pub(crate) async fn ommer_by_block_and_index(
+        &self,
+        block_id: impl Into<BlockId>,
+        index: Index,
+    ) -> EthResult<Option<RichBlock>> {
+        let block_id = block_id.into();
+        let index = usize::from(index);
+        let uncles = self.client().ommers(block_id)?.unwrap_or_default();
+        let Some(uncle_header) = uncles.get(index) else {
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/api/block.rs:34]:
```suggestion
        let Some(uncle_header) = uncles.skip(index).next() else {
```

File: crates/rpc/rpc/src/eth/api/block.rs [hunk 2]
```diff
@@ -23,6 +23,19 @@ where
...
+            return Ok(None)
+        };
+        self.block(BlockId::from(uncle_header.number), false).await
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/api/block.rs:37]: this is not correct, we need to convert he Header into a block response instead.

>  An uncle doesn't contain individual transactions.

https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_getunclebyblockhashandindex

So need to construct an RPC block from header only.

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/rpc/rpc/src/eth/api/block.rs (hunk 2 above)
Comment (fkrause98) [reply crates/rpc/rpc/src/eth/api/block.rs:37]: Does that mean that I should add an RPC type like `Block`, that lacks a `transactions` field and use it as a return type then? 😄

Comment (gakonst) [reply crates/rpc/rpc/src/eth/api/block.rs:37]: Correct it should  be the same as get block with transaction hashes only

Comment (fkrause98) [reply crates/rpc/rpc/src/eth/api/block.rs]: I've added [this struct here](https://github.com/paradigmxyz/reth/pull/1595/files#diff-42044b02ed3f935d36786aa09df65f4d748080b9ad5b93cff782badaf9a011ebR325) for the response and a few more changes. Please, let me know if there's something else to improve :slightly_smiling_face:.

File: crates/rpc/rpc/src/eth/api/block.rs (hunk 1 above)
Comment (fkrause98) [reply crates/rpc/rpc/src/eth/api/block.rs]: [Clippy suggests nth](https://rust-lang.github.io/rust-clippy/master/index.html#iter_skip_next), so I've used that instead :slightly_smiling_face: .

File: crates/rpc/rpc-types/src/eth/block.rs [hunk 3]
```diff
@@ -318,6 +319,28 @@ impl<T: Serialize> Serialize for Rich<T> {
     }
 }
 
+/// An uncle block RPC response representation.
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct OmmerBlock {
```
Comment (mattsse) [inline crates/rpc/rpc-types/src/eth/block.rs:325]: I think there's been some confusion about what to return for uncle.

we need to return the actual Block type:

e.g.

```
cast rpc eth_getUncleByBlockNumberAndIndex "0x27102" "0x0" --rpc-url "https://eth-mainnet.alchemyapi.io/v2/Lc7oIGYeL_QvInzI0Wiu_pOZZDEKBrdf" | jq
```

```json
{
  "difficulty": "0x6222410d3e2",
  "extraData": "0xd783010002844765746887676f312e342e32856c696e7578",
  "gasLimit": "0x2fefd8",
  "gasUsed": "0x0",
  "hash": "0x7ddcbe9b51cbead49b0edbb49b1a889a9cbc687cebd2cf05af43124d92472b5b",
  "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
  "miner": "0x52bc44d5378309ee2abf1539bf71de1b7d7be3b5",
  "mixHash": "0x98786cd3fc481f59562137daea90e3b7663eacbcee64f758d15e38aca31089c5",
  "nonce": "0xdb601d60e4b0146a",
  "number": "0x270ff",
  "parentHash": "0x00e1e32b60c8a8b24ad81cd9cecab49574424add31933c861bea90df2dce628b",
  "receiptsRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
  "sha3Uncles": "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
  "size": "0x21d",
  "stateRoot": "0xda97dd606da5ed4642c2fbb0c80c0cbf6951d5fd175bb258e538c9edaf6fcfab",
  "timestamp": "0x55e19b96",
  "transactionsRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
  "uncles": []
}
```

so it's the same as a regular but but some fields are default values, like bloom, gasUsed

File: crates/rpc/rpc/src/eth/api/block.rs [hunk 4]
```diff
@@ -27,14 +27,12 @@ where
...
-    ) -> EthResult<Option<RichBlock>> {
+    ) -> EthResult<Option<OmmerBlock>> {
         let block_id = block_id.into();
         let index = usize::from(index);
         let uncles = self.client().ommers(block_id)?.unwrap_or_default();
-        let Some(uncle_header) = uncles.get(index) else {
-            return Ok(None)
-        };
-        self.block(BlockId::from(uncle_header.number), false).await
+        let uncle = uncles.into_iter().skip(index).next().map(|uncle| uncle.into());
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/api/block.rs]: oh this should be equivalent to `nth()`

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/rpc/rpc-types/src/eth/block.rs (hunk 3 above)
Comment (fkrause98) [reply crates/rpc/rpc-types/src/eth/block.rs]: Oh, ok! Pushed some changes to use the original Block type, and this is what I get:
```bash
curl --location --request POST 'http://127.0.0.1:8545/' \
      --header 'Content-Type: application/json' \
      --data-raw '{
          "jsonrpc": "2.0",
          "method": "eth_getUncleByBlockNumberAndIndex",
          "params":
          [
              "0x21",
              "0x0"
          ],
          "id": "0"
  }
  ' | jq -S ".result"
```
```json
{
  "author": "0xdb312d1d6a2ccc64dd94a3892928bac82b4e8c15",
  "difficulty": "0x40888145a",
  "extraData": "0x476574682f76312e302e302d30636463373634372f6c696e75782f676f312e34",
  "gasLimit": "0x1388",
  "gasUsed": "0x0",
  "hash": "0x962b18a4fcdd215e11db01308f6818735ea1b33377a82af39cb89fa6d4178204",
  "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
  "miner": "0xdb312d1d6a2ccc64dd94a3892928bac82b4e8c15",
  "mixHash": "0x5870078c7ab1637370a603260f3b25d5f006077df2fa0f868d0970ea32496075",
  "nonce": "0x3673ed3d817b95bf",
  "number": "0x1f",
  "parentHash": "0xd1eb5ef56d30f0d773cccd3075c7fcca14ec9013db6b76344c69dbbaffaa9b8c",
  "receiptsRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
  "sha3Uncles": "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
  "size": "0x220",
  "stateRoot": "0x1624d23c380d999ac180022f08f008849ed5a404e50a0a8c704d5a536742e471",
  "timestamp": "0x55ba4322",
  "transactionsRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
  "uncles": []
}

```

It matches the response that I get from Alchemy, except from one small thing, the "author" key.

**Review by mattsse — CHANGES_REQUESTED** [review]
only a few nits, otherwise lgtm

fyi: geth marshals uncles like:

https://github.com/ethereum/go-ethereum/blob/ee530c0d5aa70d2c00ab5691a89ab431b73f8165/internal/ethapi/api.go#L814-L814

which we do now mirror.

I wonder if these changes would make the block type less usable for third party crates. But for simplicity this is ok.

File: crates/rpc/rpc-types/src/eth/block.rs [hunk 5]
```diff
@@ -56,17 +65,20 @@ pub struct Block {
...
     /// Total difficulty
-    pub total_difficulty: U256,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub total_difficulty: Option<U256>,
     /// Uncles' hashes
     pub uncles: Vec<H256>,
     /// Transactions
-    pub transactions: BlockTransactions,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub transactions: Option<BlockTransactions>,
```
[the remark this answers is not in the document]
Comment (fkrause98) [reply crates/rpc/rpc-types/src/eth/block.rs]: You're right, [changed it here](https://github.com/paradigmxyz/reth/pull/1595/files#diff-42044b02ed3f935d36786aa09df65f4d748080b9ad5b93cff782badaf9a011ebR74) :slightly_smiling_face:

File: crates/rpc/rpc-types/src/eth/block.rs [hunk 6]
```diff
@@ -56,17 +65,20 @@ pub struct Block {
     #[serde(flatten)]
     pub header: Header,
     /// Total difficulty
-    pub total_difficulty: U256,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub total_difficulty: Option<U256>,
```
[the remark this answers is not in the document]
Comment (fkrause98) [reply crates/rpc/rpc-types/src/eth/block.rs:70]: Sure! [Here its is](https://github.com/paradigmxyz/reth/pull/1595/files#diff-42044b02ed3f935d36786aa09df65f4d748080b9ad5b93cff782badaf9a011ebR67)

**Review by mattsse — APPROVED** [review]
lgtm
tysm, nice job!

**Pull request merged**